core/src/services/github/lister.rs

```diff
@@ -24,27 +24,34 @@ use crate::*;
 
 pub struct GithubLister {
     core: Arc<GithubCore>,
+
     path: String,
+    recursive: bool,
 }
 
 impl GithubLister {
-    pub fn new(core: Arc<GithubCore>, path: &str) -> Self {
+    pub fn new(core: Arc<GithubCore>, path: &str, recursive: bool) -> Self {
         Self {
             core,
 
             path: path.to_string(),
+            recursive,
         }
     }
 }
 
 impl oio::PageList for GithubLister {
     async fn next_page(&self, ctx: &mut oio::PageContext) -> Result<()> {
-        let entries = self.core.list(&self.path).await?;
+        let resp = self.core.list(&self.path).await?;
+
+        // Record whether there is a dir in the list so that we need recursive later.
+        let mut has_dir = false;
 
-        for entry in entries {
+        for entry in resp.entries {
```

Comment:
We can loop `resp.entries` first to check `has_dir` first before parse `Entry`.

Reply:
The reason for this is that if dir doesn't exist, we don't need to iterate through resp.entries.

Reply:
"iterate through resp.entries" is cheap, I care more about readability here.

Reply:
fixed